### PR TITLE
Add support for reduced precision scalars

### LINKED ISSUE #29: Half-precision scalars
Currently we are able to create reduced-precision Scalar constants via e.g. `IrBuilder::create<Double>(3.14, DataType::Half)`. In Python, we can call `fd.define_constant(3.14, dtype=DataType.Half)` and similar for `fd.define_scalar`. However, if we use them in a fusion, we get compile errors:
```c++
  auto tv0 = makeSymbolicTensor(1, DataType::Half);
  fusion->addInput(tv0);

  auto tv2 = full_like(tv0, IrBuilder::create<Double>(1.5, DataType::Half));
  fusion->addOutput(tv2);
  // C++ exception with description "false INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/ir_interface_nodes.h":111, please report a bug
  // to PyTorch. Invalid data type: __half                                                                                                   Exception raised from toString at /opt/pytorch/nvfuser/csrc/ir_interface_nodes.h:111 (most recent call first):
```
I believe there are just a few places where we need to handle printing constants and scalars to the kernel. We just need to inspect the kernels to ensure no unnecessary casts are inserted.

### PR DESCRIPTION
**Pull request by jacobhinkle**
Fixes #29

### CODE REVIEW

**Comment by jacobhinkle** [comment]
Test currently fails for input scalars (as opposed to constants). Kernels and output look like:
```c++
__global__ void kernel1(int64_t i0, Tensor<__half, 1> T2) {                                                                              
  int i32;                                                          
  i32 = ((nvfuser_index_t)threadIdx.x) + (128 * ((nvfuser_index_t)blockIdx.x));                                                          
  if ((i32 < i0)) {                                                                                                                      
    __half T5[1];                                                                                                                        
    T5[0] = 1.50000e+00;                                                                                                                 
    T2[i32]                                                                                                                              
       = T5[0];                                                                                                                          
  }                                                                                                                                      
}                                                                                                                                                                                                                       
__global__ void kernel2(int64_t i2, Tensor<__bfloat, 1> T3) {                                                                            
  int i32;                                                                                                                               
  i32 = ((nvfuser_index_t)threadIdx.x) + (128 * ((nvfuser_index_t)blockIdx.x));
  if ((i32 < i2)) {                                                 
    __bfloat T5[1];                                                                                                                      
    T5[0] = 1.7000e+00;                                             
    T3[i32]                                                         
       = T5[0];                                                       }                                                                                                                                      
}                                                                                                                
__global__ void kernel3(__half f5, int64_t i0, Tensor<__half, 1> T4) {                                                                   
  int i32;                                                                                                                               
  i32 = ((nvfuser_index_t)threadIdx.x) + (128 * ((nvfuser_index_t)blockIdx.x));
  if ((i32 < i0)) {                                                                                                                      
    __half T5[1];                                                                                                                        
    T5[0] = f5;                                                                                                                          
    T4[i32]                                                                                                                              
       = T5[0];                                                                                                                          
  }                                                                                                                                      
}                                                                                                                                        
/* output:
 1.5000                
 1.5000  
 1.5000 
 1.5000  
 1.5000                                                                                                                                  
[ CUDAHalfType{5} ]                                                                                                                      
 1.7031
 1.7031                           
 1.7031                                                                 
 1.7031                                                             
 1.7031                                                                                                                                  
[ CUDABFloat16Type{5} ]                                                                                                                  
 1638     
 1638                                                                                                                                    
 1638                                                               
 1638            
 1638          
[ CUDAHalfType{5} ]                                                 
unknown file: Failure
C++ exception with description "aten_output_tensor.allclose( fusion_output_tensor.to(aten_output_tensor.dtype()), tolerance_values.second
, tolerance_values.first, true) INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/test/test_gpu_validator.h":400, please report a bug to Py
Torch.                                                                                                                                   
                                                                    
Validation error in output 2 on line 7930 in file /opt/pytorch/nvfuser/test/test_gpu3.cpp.
  Detected abs error of: 1637
    absolute tolerance was set to 0.00390625
    and relative tolerance set to 0.0078125
*/
```
Once I get that last part to look normal, I'll mark as ready.

**Comment by jacobhinkle** [comment]
Looks like we're just missing a case here https://github.com/NVIDIA/Fuser/blob/main/csrc/executor_utils.cpp#L876-L890

Interesting that I don't see any complex cases handled there either. I will check on complex input scalars while I'm here.

File: csrc/codegen.cpp [hunk 1]
```diff
@@ -156,6 +156,21 @@ class CudaKernelGenerator : private OptOutConstDispatch {
       digits = std::numeric_limits<float>::max_digits10;
     } else if (dtype == DataType::Double) {
       digits = std::numeric_limits<double>::max_digits10;
+    } else if (dtype == DataType::Half) {
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/codegen.cpp:159]: Yeah. It is used twice so makes sense. I will move it.

**Comment by jacobhinkle** [comment]
It remains impossible to use `Scalar`s of type `ComplexFloat`, `Half`, or `BFloat16` _as inputs_ to a fusion. This was not working previously either, but I added those checks to `Fusion::addInput` which previously only checked the `Float` case. Using any of these as constants works as expected now.

**Review by csarofeen — APPROVED** [review]
LGTM

**Pull request merged**